File: spk/cli/_cmd_search.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import Any
import argparse

import structlog

import spk
from spk.io import format_ident

from . import _flags

_LOGGER = structlog.get_logger("cli")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    search_cmd = sub_parsers.add_parser(
        "search", help=_search.__doc__, description=_search.__doc__, **parser_args
    )
    _flags.add_repo_flags(search_cmd)
    search_cmd.add_argument("term", metavar="TERM", help="The search term / substring")
    search_cmd.set_defaults(func=_search)
    return search_cmd


def _search(args: argparse.Namespace) -> None:
    """Search for packages by name/substring."""

    repos = _flags.get_repos_from_repo_flags(args)

    width = max(map(len, repos.keys()))
    for repo_name, repo in repos.items():
        for name in repo.list_packages():
            if args.term in name:
                versions = list(
                    spk.api.Ident(name, spk.api.parse_version(v))
                    for v in repo.list_package_versions(name)
                )
                for v in versions:
                    print(
                        ("{: <" + str(width) + "}").format(repo_name), format_ident(v)
                    )

File: src/cli/cmd_search.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk
use anyhow::Result;
use clap::Args;

use super::{flags, CommandArgs, Run};

/// Search for packages by name/substring
#[derive(Args)]
pub struct Search {
    #[clap(flatten)]
    pub repos: flags::Repositories,

    #[clap(short, long, global = true, parse(from_occurrences))]
    pub verbose: u32,

    /// The text/substring to search for in package names
    term: String,
}

#[async_trait::async_trait]
impl Run for Search {
    async fn run(&mut self) -> Result<i32> {
        let repos = self.repos.get_repos(&["origin".to_string()]).await?;

        let width = repos
            .iter()
            .map(|(n, _)| n)
            .map(String::len)
            .max()
            .unwrap_or_default();
        let mut exit = 1;
        for (repo_name, repo) in repos.iter() {
            for name in repo.list_packages().await? {
                if !name.as_str().contains(&self.term) {
                    continue;
                }
                let mut ident = spk::api::parse_ident(&name)?;
                let versions = repo.list_package_versions(&name).await?;
                for v in versions.iter() {
                    ident.version = (**v).clone();
                    exit = 0;
                    println!("{repo_name: <width$} {}", spk::io::format_ident(&ident));
                }
            }
        }
        Ok(exit)
    }
}

impl CommandArgs for Search {
    fn get_positional_args(&self) -> Vec<String> {
        // The important positional arg for a search is the search term
        vec![self.term.clone()]
    }
}
